File: student-2/backend/Api/Models/Automation.cs
using Api.DTOs;
using Api.Extensions;

namespace Api.Models;

public abstract class Automation
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Guid StudentId { get; set; }
    public bool Enabled { get; set; }
    public bool Deleted { get; set; }
    public ICollection<AutomationRun> Runs { get; set; } = [];
    public abstract AutomationDto ToDto();
}

public class AssignmentExtensionAutomation : Automation
{
    public long? SubjectId { get; set; }
    public int BufferMinutes { get; set; }
    public AssignmentExtensionReason Reason { get; set; }
    public string FurtherDetails { get; set; } = string.Empty;

    public override AutomationDto ToDto()
    {
        return new AssignmentExtensionAutomationDto
        {
            Id = Id,
            StudentId = StudentId,
            Enabled = Enabled,
            SubjectId = SubjectId,
            BufferMinutes = BufferMinutes,
            Reason = Reason,
            FurtherDetails = FurtherDetails
        };
    }
}

public class ScheduledPostAutomation : Automation
{
    public DateTime PostTime { get; set; }
    public string ContextCode { get; set; } = string.Empty;
    public string Recipients { get; set; } = "[]";
    public string Subject { get; set; } = string.Empty;
    public string Body { get; set; } = string.Empty;
    public bool GroupConversation { get; set; }

    public override AutomationDto ToDto()
    {
        return new ScheduledPostAutomationDto
        {
            Id = Id,
            StudentId = StudentId,
            Enabled = Enabled,
            PostTime = PostTime,
            ContextCode = ContextCode,
            Recipients = RecipientJson.Deserialize(Recipients),
            Subject = Subject,
            Body = Body,
            GroupConversation = GroupConversation
        };
    }
}

public class QuizFillerAutomation : Automation
{
    public long? SubjectId { get; set; }
    public bool MultipleChoice { get; set; }
    public bool ShortAnswer { get; set; }
    public int NumberOfAttemptsRequired { get; set; }
    public bool AllowForNoTimeLimit { get; set; }

    public override AutomationDto ToDto()
    {
        return new QuizFillerAutomationDto
        {
            Id = Id,
            StudentId = StudentId,
            Enabled = Enabled,
            SubjectId = SubjectId,
            MultipleChoice = MultipleChoice,
            ShortAnswer = ShortAnswer,
            NumberOfAttemptsRequired = NumberOfAttemptsRequired,
            AllowForNoTimeLimit = AllowForNoTimeLimit
        };
    }
}

File: student-2/frontend/src/types/automation.ts
interface AutomationBase {
  id: string
  studentId: string
  enabled: boolean
}

export type AssignmentExtensionReason =
  | 'UNW'
  | 'ACL'
  | 'NMT'
  | 'FAM'
  | 'CAR'
  | 'REL'
  | 'WRK'
  | 'TEC'
  | 'BRV'
  | 'OTH'

export interface AssignmentExtensionAutomation extends AutomationBase {
  $type: 'assignmentExtension'
  subjectId: number | null
  bufferMinutes: number
  reason: AssignmentExtensionReason
  furtherDetails: string
}

export interface ScheduledPostAutomation extends AutomationBase {
  $type: 'scheduledPost'
  postTime: string
  contextCode: string
  recipients: string[]
  subject: string
  body: string
  groupConversation: boolean
}

export interface QuizFillerAutomation extends AutomationBase {
  $type: 'quizFiller'
  subjectId: number | null
  multipleChoice: boolean
  shortAnswer: boolean
  numberOfAttemptsRequired: number
  allowForNoTimeLimit: boolean
}

export type Automation =
  | AssignmentExtensionAutomation
  | ScheduledPostAutomation
  | QuizFillerAutomation
export type AutomationDiscriminator = Automation['$type']

interface SaveAutomationInputBase {
  studentId: string
  enabled: boolean
}

export interface SaveAssignmentExtensionAutomationInput extends SaveAutomationInputBase {
  $type: 'assignmentExtension'
  subjectId: number | null
  bufferMinutes: number
  reason: AssignmentExtensionReason
  furtherDetails: string
}

export interface SaveScheduledPostAutomationInput extends SaveAutomationInputBase {
  $type: 'scheduledPost'
  postTime: string
  contextCode: string
  recipients: string[]
  subject: string
  body: string
  groupConversation: boolean
}

export interface SaveQuizFillerAutomationInput extends SaveAutomationInputBase {
  $type: 'quizFiller'
  subjectId: number | null
  multipleChoice: boolean
  shortAnswer: boolean
  numberOfAttemptsRequired: number
  allowForNoTimeLimit: boolean
}

export type SaveAutomationInput =
  | SaveAssignmentExtensionAutomationInput
  | SaveScheduledPostAutomationInput
  | SaveQuizFillerAutomationInput

interface AutomationFormDataBase {
  enabled: boolean
}

export interface AssignmentExtensionAutomationFormData extends AutomationFormDataBase {
  $type: 'assignmentExtension'
  subjectId: number | null
  bufferMinutes: number
  reason: AssignmentExtensionReason
  furtherDetails: string
}

export interface ScheduledPostAutomationFormData extends AutomationFormDataBase {
  $type: 'scheduledPost'
  postTime: string
  contextCode: string
  recipients: string[]
  subject: string
  body: string
  groupConversation: boolean
}

export interface QuizFillerAutomationFormData extends AutomationFormDataBase {
  $type: 'quizFiller'
  subjectId: number | null
  multipleChoice: boolean
  shortAnswer: boolean
  numberOfAttemptsRequired: number
  allowForNoTimeLimit: boolean
}

export type AutomationFormData =
  | AssignmentExtensionAutomationFormData
  | ScheduledPostAutomationFormData
  | QuizFillerAutomationFormData

interface AutomationRunBase {
  id: string
  automationId: string
  executionTimeStamp: string
  result: 'RUN' | 'SUC' | 'FAI'
}

export interface AssignmentExtensionAutomationRun extends AutomationRunBase {
  $type: 'assignmentExtension'
  assignmentId: string
}

export interface ScheduledPostAutomationRun extends AutomationRunBase {
  $type: 'scheduledPost'
  postTime: string
  contextCode: string
  recipients: string[]
  subject: string
  body: string
  groupConversation: boolean
}

export interface QuizFillerAutomationRun extends AutomationRunBase {
  $type: 'quizFiller'
  courseId: number
  quizId: number
  quizTitle: string
  questionCount: number
}

export type AutomationRun =
  | AssignmentExtensionAutomationRun
  | ScheduledPostAutomationRun
  | QuizFillerAutomationRun